### PR TITLE
feat(SU_Accountability): Log superuser access with reason

### PR DESCRIPTION
**Pull request by maxiuyuan**
This PR can’t be merged until frontend is in place

Backend changes for Better Superuser Accountability:

Logs the category of access and reason for accessing superuser. 

Functionality: After filling out the frontend forms (future pr), the given category of access and reason are logged along with the user's id and email to google logs and will be available on Looker at a later time. When the superuser session expires, the reasons for access and orgs access will be wiped as well

PRs in this epic:
https://github.com/getsentry/sentry/pull/31982/files - Log su when during privilege org access
https://github.com/getsentry/sentry/pull/32816 - changed expiry time for privilege org access
https://github.com/getsentry/sentry/pull/32191 - disable default su access in login flow 

FE PRs: 
https://github.com/getsentry/sentry/pull/32556 - Modified Superuser and Sudo modal
https://github.com/getsentry/getsentry/pull/7146 - getsentry hook to keep Superuser access categories private

Order of merging:

First: FE changes
next: This pr as it is the base everything else merges into
The order for the other 2 doesn't matter since they are independent of one another

[notion doc for more detail of each pr with action diagram and user story](https://www.notion.so/sentry/Better-Accountability-for-Superuser-access-PRs-0854562e6b4a4a1aa7744a7f43b82a54)

#sync-getsentry

### CODE REVIEW

File: src/sentry/auth/superuser.py
Comment (dcramer) [inline src/sentry/auth/superuser.py:44]: this means a superuser session can only last 15 minutes, no?

Comment (maxiuyuan) [reply src/sentry/auth/superuser.py:44]: Yup, we want to limit the superuser session to 15 mins

Comment (dcramer) [reply src/sentry/auth/superuser.py:44]: whats driving this? its going to make _admin really hard to use

Comment (maxiuyuan) [reply src/sentry/auth/superuser.py:44]: https://www.notion.so/sentry/23Q1-Better-accountability-for-superuser-access-3a9d48bc2326467990c70f771bfa32a0

Comment (dcramer) [reply src/sentry/auth/superuser.py:44]: that makes sense, and the premise of this makes sense, but i think theres a large delta in accessing customer data (e.g. viewing an orgs data) and utilizing _admin which is accessing sentry data (we have definitions for this buried in notion). if there were a way to say "you can access customer's data aka events for 15 minutes when you put in a reason" and it kept the 4 hour timer on _admin/general management stuff i think that'd be a great win

however with this I see something like: im in _admin, i click around, im 12 minutes into my session and uploading a file for a policy change, and the session has expired.

Comment (maxiuyuan) [reply src/sentry/auth/superuser.py:44]: hmm good point, ill bring it up for discussion

Comment (maxiuyuan) [reply src/sentry/auth/superuser.py:44]: after discussion, we plan to keep it at 4 hours unless you're accessing another org, in which then it will be 15 mins

File: src/sentry/auth/superuser.py
Comment (JoshFerge) [inline src/sentry/auth/superuser.py:47]: is "changed org" the right name here? what do we think about something like "PRIVILEGED_ORG_ACCESS" instead?

File: src/sentry/auth/superuser.py
Comment (JoshFerge) [inline src/sentry/auth/superuser.py:100]: can we put this in a separate PR? seems like a separate concern of our logging we want to do.

Comment (maxiuyuan) [reply src/sentry/auth/superuser.py:100]: yup will do!

File: src/sentry/auth/superuser.py
Comment (JoshFerge) [inline src/sentry/auth/superuser.py:323]: what's the purpose of this `except` block?

File: src/sentry/auth/superuser.py
Comment (JoshFerge) [inline src/sentry/auth/superuser.py:68]: i would rename these for clarity:
`SuperuserAccessCategory`
`SuperuserReason`

File: src/sentry/auth/superuser.py
Comment (maxiuyuan) [reply src/sentry/auth/superuser.py:323]: I put that incase `json.loads(request.body)` errored out, i haven't seen it happen so i'm not sure if it was needed

File: src/sentry/api/endpoints/auth_index.py
Comment (maxiuyuan) [inline src/sentry/api/endpoints/auth_index.py:113]: @JoshFerge @leedongwei For the users without passwords trying to enable active su, do we want to make sure they were last verified within a certain time (15 mins before) otherwise we ask them to re-authenticate or is the default sso session time (20 hours?) good enough?

File: src/sentry/auth/superuser.py
Comment (JoshFerge) [inline src/sentry/auth/superuser.py:47]: did we all agree on this change? if more discussion needs to be had, may be worth putting this change in a separate PR.

Comment (maxiuyuan) [reply src/sentry/auth/superuser.py:47]: No, there wasn't any discussion. Although i feel like 15 mins would be better if we plan to limit privileged access 15 mins. What do u think? Ill ask danny as well and ill [move it](https://github.com/getsentry/sentry/pull/32104) depending on what we decide on

File: src/sentry/auth/superuser.py
Comment (RyanSkonnord) [inline src/sentry/auth/superuser.py:312]: I'd suggest extracting this new code block out to a helper method. In addition to keeping `set_logged_in` shorter overall, you could shrink the `try` block as:

```python
    def _...(...):
        try:
            su_access_json = json.loads(request.body)
        except json.JSONDecodeError:
            return

        if "superuserAccessCategory" in su_access_json:
        # rest of method...
```

File: src/sentry/utils/auth.py
Comment (RyanSkonnord) [inline src/sentry/utils/auth.py:240]: Nit: delete `#`?

File: src/sentry/auth/superuser.py
Comment (JoshFerge) [inline src/sentry/auth/superuser.py:65]: did we want this to be a `ChoiceField`, with the reasons populated dynamically?

File: src/sentry/auth/superuser.py
Comment (JoshFerge) [inline src/sentry/auth/superuser.py:303]: let's rename the var `token` to `superuser_session_id`

File: src/sentry/auth/superuser.py
Comment (maxiuyuan) [reply src/sentry/auth/superuser.py:65]: yup i'll make the changes once i set the reasons

File: src/sentry/auth/superuser.py
Comment (maxiuyuan) [reply src/sentry/auth/superuser.py:303]: do we want to change all instances of `token` into `superuser_session_id` or just within this method?

Comment (JoshFerge) [reply src/sentry/auth/superuser.py:303]: if its used in a lot of places, then makes sense to log it as `superuser_token_id`

File: tests/sentry/api/endpoints/test_auth_index.py
Comment (JoshFerge) [inline tests/sentry/api/endpoints/test_auth_index.py:133]: what happens if the user sends a password along with the superuser category/reason?

Comment (maxiuyuan) [reply tests/sentry/api/endpoints/test_auth_index.py:133]: removing the password was a copy and paste mistake on my part, this actually changes the behaviour of this test case so i will fix

File: src/sentry/api/endpoints/auth_index.py
Comment (JoshFerge) [inline src/sentry/api/endpoints/auth_index.py:120]: what if the user has multiple AuthIdentitys?

File: src/sentry/api/endpoints/auth_index.py
Comment (JoshFerge) [inline src/sentry/api/endpoints/auth_index.py:123]: what if a superuser is just trying to use `sudo` and not `superuser`? should we make it so that they may `sudo` without becoming `superuser`?

File: src/sentry/api/endpoints/auth_index.py
Comment (JoshFerge) [inline src/sentry/api/endpoints/auth_index.py:115]: won't this completely error out in the non superuser case?

File: src/sentry/api/endpoints/auth_index.py
Comment (JoshFerge) [inline src/sentry/api/endpoints/auth_index.py:187]: what if a non-superuser accesses the superuser modal? should you still check that the user is superuser here?

File: src/sentry/auth/superuser.py
Comment (JoshFerge) [inline src/sentry/auth/superuser.py:306]: worth adding a comment as to why you have to use json.loads here.

File: src/sentry/conf/server.py
Comment (JoshFerge) [inline src/sentry/conf/server.py:2515]: didn't we want to make this private? (i.e., not in the `sentry` repo)

**Review by JoshFerge — CHANGES_REQUESTED** [review]
The non-superuser AuthIdentity only case looks completely broken to me. I would refer back to your flowchart and ensure we have an integration test for each final state.

File: src/sentry/api/endpoints/auth_index.py
Comment (maxiuyuan) [reply src/sentry/api/endpoints/auth_index.py:120]: this would be the case if the user has an authidentity in multiple orgs right? i want to redirect them to the sso login of the org they're currently in, ill look for another way to do this

File: src/sentry/api/endpoints/auth_index.py
Comment (maxiuyuan) [reply src/sentry/api/endpoints/auth_index.py:115]: yeah, i need to fix this and get the non superuser's org

File: src/sentry/api/endpoints/auth_index.py
Comment (maxiuyuan) [reply src/sentry/api/endpoints/auth_index.py:187]: from the frontend it shouldn't possible to open the superuser if the user isn't a non-superuser, but the line right under this checks if the user is a superuser and if they're not an active superuser

File: src/sentry/api/endpoints/auth_index.py
Comment (maxiuyuan) [reply src/sentry/api/endpoints/auth_index.py:123]: this shouldn't affect if they're trying to sudo or superuser, this just redirects them to their org's sso if they don't have a valid sso session

Comment (JoshFerge) [reply src/sentry/api/endpoints/auth_index.py:123]: what if they're not in their superuser org? 🙃  i think we've found more test cases to add lol

File: src/sentry/api/endpoints/auth_index.py
Comment (JoshFerge) [inline src/sentry/api/endpoints/auth_index.py:51]: nit: `_needs_redirect`

File: src/sentry/auth/superuser.py
Comment (JoshFerge) [inline src/sentry/auth/superuser.py:315]: where does AtttributeError come from? how can this happen?

File: src/sentry/auth/superuser.py
Comment (JoshFerge) [inline src/sentry/auth/superuser.py:316]: nit: if the `su_access_json` above is to be a dict once loaded, its also a good idea to make this a dict

File: src/sentry/api/endpoints/auth_index.py
Comment (maxiuyuan) [reply src/sentry/api/endpoints/auth_index.py:176]: I paired with danny on this one and the "real request object that Django knows about" was confusing, for the django framework, it seems `Django HttpRequest` is needed instead of the `django rest framework Request`

File: src/sentry/auth/superuser.py
Comment (JoshFerge) [inline src/sentry/auth/superuser.py:90]: how can this `AttributeError` happen?

File: src/sentry/conf/server.py
Comment (JoshFerge) [reply src/sentry/conf/server.py:2520]: @maxiuyuan why did you resolve this conversation?

File: src/sentry/auth/superuser.py
Comment (JoshFerge) [inline src/sentry/auth/superuser.py:298]: if we have an option in settings that is default off, i don't think you need to check `is_self_hosted` here.

File: src/sentry/api/endpoints/auth_index.py
Comment (JoshFerge) [inline src/sentry/api/endpoints/auth_index.py:129]: i'm confused, doesn't this break for non superusers?

Comment (JoshFerge) [reply src/sentry/api/endpoints/auth_index.py:129]: do we have tests for non superuser sso?

**Review by JoshFerge — CHANGES_REQUESTED** [review]
nits + questions around non superuser sso requests to this endpoint still.

File: src/sentry/conf/server.py
Comment (maxiuyuan) [reply src/sentry/conf/server.py:2520]: oops, i removed this on my working branch and not this one, will remove this too

File: src/sentry/auth/superuser.py
Comment (maxiuyuan) [reply src/sentry/auth/superuser.py:316]: yup, i realized this in the [following pr](https://github.com/getsentry/sentry/blob/18459105cdbe444582d39a9065aed163f86ce5ee/src/sentry/auth/superuser.py#L316), ill make the change here as well

File: src/sentry/auth/superuser.py
Comment (maxiuyuan) [reply src/sentry/auth/superuser.py:315]: i'll make this. more specific, this comes from the if there is an error from `json.loads(request.body)`

File: src/sentry/auth/superuser.py
Comment (maxiuyuan) [reply src/sentry/auth/superuser.py:90]: i'll remove this, this won't happen if the setting is set

File: src/sentry/api/endpoints/auth_index.py
Comment (maxiuyuan) [reply src/sentry/api/endpoints/auth_index.py:129]: this test will test this behaviour `test_no_password_no_u2f`. this will just set `has_valid_sso_session` to false which will `return self.respond(validator.errors, status=status.HTTP_400_BAD_REQUEST)` for any non superuser

File: src/sentry/auth/superuser.py
Comment (JoshFerge) [inline src/sentry/auth/superuser.py:310]: did you want to look into using request.POST instead?

File: src/sentry/api/endpoints/auth_index.py
Comment (JoshFerge) [reply src/sentry/api/endpoints/auth_index.py:129]: right, is that intended behavior? what about people using this endpoint to `sudo`?

File: src/sentry/auth/superuser.py
Comment (JoshFerge) [reply src/sentry/auth/superuser.py:313]: you can put both the exceptions on the same line:
`except (json.JSONDecodeError, AttributeError)`.

Although if it is invalid json, wouldn't we want to return some sort of error to the client instead?

File: src/sentry/auth/superuser.py
Comment (maxiuyuan) [reply src/sentry/auth/superuser.py:310]: i looked into it, once the request reaches here, `request.POST` contains an empty query dict `<QueryDict: {}>`

File: src/sentry/api/endpoints/auth_index.py
Comment (JoshFerge) [reply src/sentry/api/endpoints/auth_index.py:129]: via the flowchart you made, does this use case still work?
Screen Shot 2022-03-30 at 1 04 51 PM

Comment (maxiuyuan) [reply src/sentry/api/endpoints/auth_index.py:129]: it is the intended behavior, people using sudo shouldn't hit this without having `validator.is_valid()` being true, so the 400 wouldn't be returned. If they somehow hit this through sudo without a password or u2f, a 400 should be returned

Comment (maxiuyuan) [reply src/sentry/api/endpoints/auth_index.py:129]: the flowchart still works since the redirect is now on the fe if the user has no password

Comment (maxiuyuan) [reply src/sentry/api/endpoints/auth_index.py:129]: image

File: src/sentry/api/endpoints/auth_index.py
Comment (JoshFerge) [inline src/sentry/api/endpoints/auth_index.py:125]: should you add `isSuperUserModal` to the `AuthVerifyValidator` and then access it via that?

File: src/sentry/api/endpoints/auth_index.py
Comment (JoshFerge) [inline src/sentry/api/endpoints/auth_index.py:180]: should this block be in this `try`?

File: src/sentry/api/endpoints/auth_index.py
Comment (JoshFerge) [inline src/sentry/api/endpoints/auth_index.py:109]: let's add another case here to return False here if there's no Superuser.org_id and authenticated is None, just to be safe.

File: src/sentry/api/endpoints/auth_index.py
Comment (JoshFerge) [inline src/sentry/api/endpoints/auth_index.py:50]: since the helper methods we added don't use self, you can make it a static method

**Review by JoshFerge — APPROVED** [review]
looks good. try and get one more +1 before merging, and lets deploy this to staging to test before merging as well.

File: src/sentry/api/endpoints/auth_index.py
Comment (nhsiehgit) [inline src/sentry/api/endpoints/auth_index.py:45]: Is there a difference between `redirect = ""` and `redirect = None`?

File: src/sentry/api/endpoints/auth_index.py
Comment (nhsiehgit) [inline src/sentry/api/endpoints/auth_index.py:109]: 🤔 
Is this just setting authenticated to true if the user is a superuser?

File: src/sentry/utils/auth.py
Comment (nhsiehgit) [inline src/sentry/utils/auth.py:248]: Should we type out the rest of these attrs too?

**Review by nhsiehgit — COMMENTED** [review]
few questions but don't think i have enough context to give an approval or not

File: src/sentry/api/endpoints/auth_index.py
Comment (maxiuyuan) [reply src/sentry/api/endpoints/auth_index.py:109]: This is a really weird case where if the user is trying to get superuser but they don't have a password or u2f device, the only way to make sure they're authenticated is to check if they have a valid sso session. We'd like to authenticate them by verifying their password or u2f so we would set authenticated to the result of that prior to this. If they don't have either of those however, we set authenticated to true if they have a valid sso session

File: src/sentry/api/endpoints/auth_index.py
Comment (maxiuyuan) [reply src/sentry/api/endpoints/auth_index.py:45]: i'm actually not too sure, this was code i refactored into a helper method but i assume its the same. Will test it out and make it one or the other for consistency

File: src/sentry/utils/auth.py
Comment (maxiuyuan) [reply src/sentry/utils/auth.py:248]: I typed request since i wanted to distinguish request as a HttpRequest over the alternative Request class. I wouldn't mind typing out the rest of those attrs tho

**Review by nhsiehgit — APPROVED** [review]
:rubber_stamp:

File: src/sentry/api/endpoints/auth_index.py
Comment (maxiuyuan) [reply src/sentry/api/endpoints/auth_index.py:45]: theres not difference it seems. if redirect is set to `""`, then `is_safe_url` will return false and redirect will be set to `None`. I'll change the default to `None` to begin with

File: src/sentry/api/endpoints/auth_index.py
Comment (RyanSkonnord) [inline src/sentry/api/endpoints/auth_index.py:95]: Does this need attention before shipping?

File: src/sentry/auth/superuser.py
Comment (RyanSkonnord) [inline src/sentry/auth/superuser.py:312]: Since these lines duplicate lines 341-350, I would pull the common code into a helper function. You could define the helper function right after `token = ...` above, nested inside `set_logged_in`, so that you can refer to variables in scope and not pass any arguments.

**Review by RyanSkonnord — APPROVED** [review]
Just a few small notes. Looks good to me.

File: src/sentry/api/endpoints/auth_index.py
Comment (maxiuyuan) [reply src/sentry/api/endpoints/auth_index.py:95]: nope!

**Pull request merged**
